pkg/liquidity-mining/contracts/gauges/StakelessGaugeController.sol

```diff
@@ -0,0 +1,223 @@
+import "@balancer-labs/v2-interfaces/contracts/liquidity-mining/IGaugeController.sol";
+import "@balancer-labs/v2-interfaces/contracts/liquidity-mining/IStakelessGauge.sol";
+import "@balancer-labs/v2-interfaces/contracts/liquidity-mining/IStakelessGaugeController.sol";
+
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/Address.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/EnumerableSet.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/ReentrancyGuard.sol";
+
+import "./arbitrum/ArbitrumRootGauge.sol";
+
+/**
+ * @title Stakeless Gauge Controller
+ * @notice Implements IStakelessGaugeController; refer to it for API documentation.
+ */
+contract StakelessGaugeController is IStakelessGaugeController, ReentrancyGuard {
+    using EnumerableSet for EnumerableSet.AddressSet;
+
+    mapping(IGaugeAdder.GaugeType => EnumerableSet.AddressSet) private _gauges;
+    IAuthorizerAdaptor private immutable _authorizerAdaptor;
+    IGaugeController private immutable _gaugeController;
+    IGaugeAdder private immutable _gaugeAdder;
+
+    constructor(
+        IAuthorizerAdaptor authorizerAdaptor,
+        IGaugeController gaugeController,
+        IGaugeAdder gaugeAdder
+    ) {
+        _authorizerAdaptor = authorizerAdaptor;
+        _gaugeController = gaugeController;
+        _gaugeAdder = gaugeAdder;
+    }
+
+    modifier gaugeTypeChecked(IGaugeAdder.GaugeType gaugeType) {
+        require(_isValidGaugeType(gaugeType), "Unsupported gauge type");
+        _;
+    }
+
+    /**
+     * @dev See IStakelessGaugeController#addGauges.
+     */
+    function addGauges(IGaugeAdder.GaugeType gaugeType, IStakelessGauge[] calldata gauges)
+        external
+        override
+        gaugeTypeChecked(gaugeType)
+    {
+        EnumerableSet.AddressSet storage gaugesForType = _gauges[gaugeType];
+
+        for (uint256 i = 0; i < gauges.length; i++) {
+            IStakelessGauge gauge = gauges[i];
+            require(
+                _gaugeAdder.isGaugeFromValidFactory(address(gauge), gaugeType),
+                "Gauge does not come from valid factory"
+            );
+            require(_gaugeController.gauge_exists(address(gauge)), "Gauge does not exist in controller");
+            require(!gauge.is_killed(), "Gauge was killed");
+            require(gaugesForType.add(address(gauge)), "Gauge already present");
+        }
+        emit IStakelessGaugeController.GaugesAdded(gaugeType, gauges);
+    }
+
+    /**
+     * @dev See IStakelessGaugeController#removeGauges.
+     */
+    function removeGauges(IGaugeAdder.GaugeType gaugeType, IStakelessGauge[] calldata gauges)
+        external
+        override
+        gaugeTypeChecked(gaugeType)
+    {
+        EnumerableSet.AddressSet storage gaugesForType = _gauges[gaugeType];
+
+        for (uint256 i = 0; i < gauges.length; i++) {
+            IStakelessGauge gauge = gauges[i];
+            require(gauge.is_killed(), "Gauge was not killed");
```

Comment:
I'd add a comment here mentioning we're not checking if the gauge comes from the factory or if it exists in the controller because we've already done that (as we're checking the gauge was added to our list), and gauges cannot be removed from either factory nor controller. Therefore, the only way to get a gauge to be removed here is to kill it.

Reply:
Done.